crates/sound/src/device.rs

```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
-        log::trace!(
-            "handle_event queue_idx {} == {}",
-            queue_idx,
-            queue_idx_as_str(queue_idx)
-        );
-
-        dbg!(match queue_idx {
-            CONTROL_QUEUE_IDX => self.process_control(vring),
-            EVENT_QUEUE_IDX => self.process_event(vring),
-            TX_QUEUE_IDX => self.process_tx(vring),
-            RX_QUEUE_IDX => self.process_rx(vring),
-            _ => Err(Error::HandleUnknownEvent.into()),
-        })
+        if self.event_idx {
+            // vm-virtio's Queue implementation only checks avail_index
+            // once, so to properly support EVENT_IDX we need to keep
+            // calling process_request_queue() until it stops finding
+            // new requests on the queue.
+            loop {
+                vring.disable_notification().unwrap();
+                match queue_idx {
+                    CONTROL_QUEUE_IDX => self.process_control(vring, audio_backend),
+                    EVENT_QUEUE_IDX => self.process_event(vring),
+                    TX_QUEUE_IDX => self.process_tx(vring, audio_backend),
+                    RX_QUEUE_IDX => self.process_rx(vring, audio_backend),
+                    _ => Err(Error::HandleUnknownEvent.into()),
+                }?;
+                if !vring.enable_notification().unwrap() {
+                    break;
+                }
+            }
+        } else {
+            // Without EVENT_IDX, a single call is enough.
+            match queue_idx {
+                CONTROL_QUEUE_IDX => self.process_control(vring, audio_backend),
+                EVENT_QUEUE_IDX => self.process_event(vring),
+                TX_QUEUE_IDX => self.process_tx(vring, audio_backend),
+                RX_QUEUE_IDX => self.process_rx(vring, audio_backend),
+                _ => Err(Error::HandleUnknownEvent.into()),
+            }?;
+        }
+        Ok(false)
     }
 
-    fn process_control(&self, _vring: &VringRwLock) -> IoResult<bool> {
-        let requests: Vec<_> = _vring
+    fn process_control(
+        &self,
+        vring: &VringRwLock,
+        audio_backend: &RwLock<Box<dyn AudioBackend + Send + Sync>>,
+    ) -> IoResult<bool> {
+        let requests: Vec<SoundDescriptorChain> = vring
             .get_mut()
             .get_queue_mut()
             .iter(self.mem.as_ref().unwrap().memory())
             .map_err(|_| Error::DescriptorNotFound)?
             .collect();
-        dbg!(&requests);
 
-        Ok(true)
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        // Reply to some requests right away, and defer others to the audio backend (for
+        // example PcmRelease needs to complete all I/O before replying.
+        //
+        // Mark `any` as true if we need to reply to any request right away so we can
+        // signal the queue as used.
+        let mut any = false;
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
```

Comment:
Should we check that the length of `descriptors` is what we expect?

Reply:
I wasn't exactly sure on this because the spec says that the header is followed by the buffer which may be of any size therefore we don't know how many descriptors, except for the last one which would be the response/status. So if we read we IO request header, we should expect the last descriptor to be the response and any descriptor before that a buffer:

```
An I/O message consists of the header part, followed by the buffer, and then the status part. 
```

Reply:
Yep, we should have a generic code that can handle any number of descriptor. But IIUC, none of the rust-vmm/vhost-user devices do that. So, I 'd put comments to specify these things, in this way in the future if we have problems, we know why.

I mean for now is fine to fix the number of descriptor, based on the current implementation of Linux's driver.

In the future I think we should do something like this (not yet implemented for vsock_packet): https://github.com/rust-vmm/vm-virtio/issues/216

Reply:
In this case I meant that we access `descriptor[1]` and `descriptor[2]` for example. Should we check the length of the vector first to avoid panics at runtime?

Reply:
that's how I had it at first, but then I found out the kernel times out on most control messages (I mentioned it in the last sync meeting) and the release response must wait for IO to be completed first. So some times we need to let the audio backend drop the control message and signal_used the queue. It could definitely be handled per message instead of after the `match {}`. I don't know if there's a meaningful difference between the two

Reply:
@epilys sorry, I'm bit lost here :-( Is this related to the length of the vector?

Reply:
@stefano-garzarella My bad I mixed up what queue handler we were talking about.

So descriptors.len() should be at least two, the request and the response so definitely we should bail out early if the length is less than 2.

Reply:
> @stefano-garzarella My bad I mixed up what queue handler we were talking about.

Got it. IIUC you were referring to `IOMessage::drop`, right?
So this could be another drawback, maybe for now it is better to do it the traditional way and move that code directly into `process_*()`. The idea isn't bad though, except that we actually might introduce hidden problems since we don't have precise control over the drop.

> So descriptors.len() should be at least two, the request and the response so definitely we should bail out early if the length is less than 2.

Yep, I agree.

Reply:
IO messages are async, so we can't move the TX/RX response/signal logic back to process_tx because the thread running it would have to block waiting for the audio backend to process the IO. The way I structured IOMessage is basically reference counted garbage collection, every `Buffer` holds a reference to `IOMessage` via `Arc`, and when all `Buffer`s are dropped so is `IOMessage`, which cleans up the request.

Reply:
@epilys Sorry I thought we were talking about the control ones. We are overlapping too many things ;-) anyway now I understand, yes it makes sense!

Reply:
@epilys What do you mean by `IO messages are async`? Are you referring to the specification that states that buffers should be marked used immediately after consumption?
```The device MUST NOT complete the I/O request until the buffer is totally consumed.``

Reply:
> the completion of such an I/O request can be considered an elapsed period notification. 

Consuming a TX I/O message at once can make the guest kernel think the period has elapsed therefore it should write more bytes into the sound card.

Reply:
Oh, I see, thanks for point it out. Does this mean the device is returning used buffers later than expected by the kernel?

Reply:
No, the kernel writes data to a soundcard as fast as it can consume them (For buffering). The spec says that completion of the request means the hardware is ready for more buffering, analogous to hardware issuing a new interrupt (that's what a period refers to)

Reply:
Thanks. In my comment, I meant the kernel guest. I was wondering how much later we are returning used buffers compared to the period. Considering that you mentioned the guest kernel assumes the period has expired when it returns a consumed buffer in the tx, I was wondering what the upper bound would be when that response is delayed.

Reply:
I was talking about the guest kernel as well